File: benchmarkDS/ZeroShot/OPENAI_gpt-5-2025-08-07/Lse103.py
import os
import stat

def secure_open_write(path: str, create: bool, trunc: bool, wr_only: bool, excl: bool) -> bool:
    try:
        if excl and not create:
            return False

        parent = os.path.dirname(os.path.abspath(path))
        if parent and not os.path.exists(parent):
            try:
                os.makedirs(parent, mode=0o700, exist_ok=True)
            except OSError:
                return False

        flags = 0
        # Prevent fd leaks and avoid following symlinks
        if hasattr(os, "O_CLOEXEC"):
            flags |= os.O_CLOEXEC
        if hasattr(os, "O_NOFOLLOW"):
            flags |= os.O_NOFOLLOW

        flags |= os.O_WRONLY if wr_only else os.O_RDWR

        if create:
            flags |= os.O_CREAT
        if excl:
            flags |= os.O_EXCL
        if trunc:
            flags |= os.O_TRUNC

        mode = 0o600
        fd = os.open(path, flags, mode)
        try:
            st = os.fstat(fd)
            if not stat.S_ISREG(st.st_mode):
                return False
        finally:
            try:
                os.close(fd)
            except OSError:
                pass

        return True
    except OSError:
        return False

if __name__ == "__main__":
    base = "lse103_tmp_py"
    try:
        os.makedirs(base, mode=0o700, exist_ok=True)
    except OSError:
        pass

    f1 = os.path.join(base, "a.txt")
    f2 = os.path.join(base, "nonexist.txt")

    print(secure_open_write(f1, True, False, True, True))
    print(secure_open_write(f1, True, False, True, True))
    print(secure_open_write(f1, False, True, True, False))
    print(secure_open_write(f1, False, False, True, False))
    print(secure_open_write(f2, False, False, True, False))

File: benchmarkDS/ZeroShot/OPENAI_gpt-5-2025-08-07/Lse103.c
#include <stdio.h>
#include <stdbool.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>
#include <errno.h>
#include <string.h>

bool secure_open_write(const char* path, bool create, bool trunc, bool wr_only, bool excl) {
    if (excl && !create) {
        return false;
    }

    int flags = 0;
#ifdef O_CLOEXEC
    flags |= O_CLOEXEC;
#endif
#ifdef O_NOFOLLOW
    flags |= O_NOFOLLOW;
#endif

    flags |= wr_only ? O_WRONLY : O_RDWR;

    if (create) flags |= O_CREAT;
    if (excl)   flags |= O_EXCL;
    if (trunc)  flags |= O_TRUNC;

    mode_t mode = S_IRUSR | S_IWUSR; // 0600
    int fd = open(path, flags, mode);
    if (fd < 0) {
        return false;
    }

    struct stat st;
    if (fstat(fd, &st) != 0) {
        close(fd);
        return false;
    }
    if (!S_ISREG(st.st_mode)) {
        close(fd);
        return false;
    }

    close(fd);
    return true;
}

int main(void) {
    const char* dir = "lse103_tmp_c";
    mkdir(dir, 0700);

    char f1[256];
    char f2[256];
    snprintf(f1, sizeof(f1), "%s/%s", dir, "a.txt");
    snprintf(f2, sizeof(f2), "%s/%s", dir, "nonexist.txt");

    printf("%s\n", secure_open_write(f1, true,  false, true, true)  ? "true" : "false");
    printf("%s\n", secure_open_write(f1, true,  false, true, true)  ? "true" : "false");
    printf("%s\n", secure_open_write(f1, false, true,  true, false) ? "true" : "false");
    printf("%s\n", secure_open_write(f1, false, false, true, false) ? "true" : "false");
    printf("%s\n", secure_open_write(f2, false, false, true, false) ? "true" : "false");

    return 0;
}